linkerd/proxy/discover/src/from_resolve.rs

```diff
@@ -111,50 +113,70 @@ where
-                        trace!(new = ?active, old = ?this.active, "Reset");
-                        for addr in this.active.iter() {
-                            // If the old addr is not in the new set, remove it.
-                            if !active.contains(addr) {
-                                trace!(%addr, "Scheduling removal");
-                                this.pending.push_back(Change::Remove(*addr));
-                            } else {
-                                trace!(%addr, "Unchanged");
-                            }
-                        }
-                        for (addr, endpoint) in endpoints.into_iter() {
-                            if !this.active.contains(&addr) {
-                                trace!(%addr, "Scheduling addition");
-                                this.pending
-                                    .push_back(Change::Insert(addr, endpoint.clone()));
-                            }
+            let update = match ready!(this.resolution.try_poll_next(cx)) {
+                Some(update) => update?,
+                None => return Poll::Ready(None),
+            };
+
+            match update {
+                Update::Reset(endpoints) => {
+                    let new_active = endpoints.into_iter().collect::<IndexMap<_, _>>();
+                    trace!(new = ?new_active, old = ?this.active, "Reset");
+
+                    for addr in this.active.keys() {
+                        // If the old addr is not in the new set, remove it.
+                        if !new_active.contains_key(addr) {
+                            trace!(%addr, "Scheduling removal");
+                            this.pending.push_back(Change::Remove(*addr));
+                        } else {
+                            trace!(%addr, "Unchanged");
                         }
-                        *this.active = active;
                     }
-                    Update::Add(endpoints) => {
-                        for (addr, endpoint) in endpoints.into_iter() {
+
+                    for (addr, endpoint) in new_active.iter() {
+                        if this.active.get(addr) != Some(endpoint) {
                             trace!(%addr, "Scheduling addition");
-                            this.active.insert(addr);
-                            this.pending.push_back(Change::Insert(addr, endpoint));
+                            this.pending
+                                .push_back(Change::Insert(*addr, endpoint.clone()));
                         }
                     }
-                    Update::Remove(addrs) => {
-                        for addr in addrs.into_iter() {
-                            if this.active.remove(&addr) {
-                                trace!(%addr, "Scheduling removal");
-                                this.pending.push_back(Change::Remove(addr));
+
+                    *this.active = new_active;
```

Comment:
spent a minute trying to figure out if there was a nice way to do this without creating a new map and dropping the old one...but i think this approach is probably the best, so :+1: